### PR TITLE
Add pipeline to publish docker container and release

### PR DESCRIPTION
**Pull request by Cisien**
on tag, this pipeline will build a docker container for common architectures (amd64, 32-bit arm, 64-bit arm) and zip up the repo as a .zip file and a .tar.gz file

the zip step is arguably not any different from the default source code zip files that are attached to a release by default, but may be useful placeholders for any kind of build steps that produce os specific packages.

How to use:
As you push changes, or new features, update CHANGELOG.md for the version that's under development, or under the Unreleased header. At some point before creating and pushing a new tag, push this changelog with the next version number as a section with all of the changes for that release. The contents of this header become the release notes.

edit CHANGELOG.md
`git add CHANGELOG.md`
`git commit -m "whatevs" `
`git push`

`git tag v1.2.3`
`git push origin v1.2.3`

### CODE REVIEW

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [inline .github/workflows/build-and-publish.yml:17]: I assume `packages` write permission is for publishing releases?

What is `contents` write permission used for?

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [inline .github/workflows/build-and-publish.yml:16]: Building on a moving target could break the build, even if we change nothing.

I'd rather target a smaller, stable image. For example `debian:11`.

What do you think?

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [inline .github/workflows/build-and-publish.yml:8]: If I understand this right, we are triggering the build process on (I assume) merged pull requests? Or on all pull requests targeting master?

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [inline .github/workflows/build-and-publish.yml:25]: We shouldn't strictly need this, but I guess it doesn't hurt.

Since we got the transwarp dependency listed in `requirements.txt` now, the submodule has become obsolete. I think we'll remove it over time.

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [inline .github/workflows/build-and-publish.yml:28]: Just for my understanding, can you explain what this line does? Why do we need to limit login to `push` actions?

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [inline .github/workflows/build-and-publish.yml:36]: I assume we need qemu for the docker build? Or?

Also, it looks like `setup-qemu-action` takes an optional list of platforms to target. Can we make the build process more light-weight by restricting the platforms to the ones we actually use?

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [inline .github/workflows/build-and-publish.yml:72]: I'd strongly prefer we use a descriptive filename. We're going to end up with a hundred files all called `ankerctl.(zip|tar.gz`.

I'm sure there's an environment var that roughly correlates to something like `git describe`?

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [inline .github/workflows/build-and-publish.yml:80]: Same here, obviously

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [inline .github/workflows/build-and-publish.yml:90]: ..and same here

Unless the build environment is completely cleared between each run, in which case we might be able to just say `release/ankerctl-*`?

**Review by chrivers — COMMENTED** [review]
Awesome work - thanks for your contribution :)

It'll be really exciting to get an automatic build system going.

Also, I'd be very nice to have a way for the tool itself to know its own version. Do you have any suggestions? (doesn't have to be in this PR)

File: .github/workflows/build-and-publish.yml
Comment (spuder) [reply .github/workflows/build-and-publish.yml:16]: This is an interesting one. I’m a hard proponent of always pinning versions, yet GitHub actions runs-on stanza might be worth an exception for the following reasons:

1. All actions documentation seems to recommend using either ‘ubuntu-latest’, ‘windows-latest’ or macos-latest’. I’ve never seen an exception to this outside of some nuanced osx builds targeting old versions. Here are the docs: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources
2. the runs on stanza isn’t referencing the official ubuntu container like you  might expect. It is actually a blessed image from GitHub. See this quote 

> Note: The -latest runner images are the latest stable images that GitHub provides, and might not be the most recent version of the operating system available from the operating system vendor.

3. Anecdotally I’ve been using actions since 2018 and I’ve never seen a build break from ubuntu-latest.

File: .github/workflows/build-and-publish.yml
Comment (Cisien) [reply .github/workflows/build-and-publish.yml:8]: any pull request targeting master -- this is a PR verification step mostly. Any actual publishing steps are skipped for these builds.

File: .github/workflows/build-and-publish.yml
Comment (Cisien) [reply .github/workflows/build-and-publish.yml:16]: likewise, I've been using ubuntu-latest since the dawn of actions and haven't had any issues from it. Here's the repo that covers the runner images.
https://github.com/actions/runner-images

latest is currently ubuntu-22.04. For linux runners, we have a choice between latest (2204), 2204 pinned, and 2004

File: .github/workflows/build-and-publish.yml
Comment (Cisien) [reply .github/workflows/build-and-publish.yml:17]: publishing releases, a setting on the project needs to be flipped as well.

Under Project settings, Actions, General:
image

File: .github/workflows/build-and-publish.yml
Comment (Cisien) [reply .github/workflows/build-and-publish.yml:28]: the `event_name` is `push` when a new tag is pushed (or technically any commit, but we're only looking at tags). For builds triggered from a pull request, the `event_name` is `pull_request` iirc.

This lets us skip publishing-related steps for builds that will never publish artifacts.

File: .github/workflows/build-and-publish.yml
Comment (Cisien) [reply .github/workflows/build-and-publish.yml:36]: The qemu step is used by buildx to build docker images for other platforms. We are targeting amd64, 32 bit arm, and 64 bit arm in the docker build step. I don't think targeting platforms here will save us much since this task only takes a few seconds to run.

File: .github/workflows/build-and-publish.yml
Comment (Cisien) [reply .github/workflows/build-and-publish.yml:72]: you're thinking something like `ankerctl-v1.0.0.(zip|tar.gz)`? I'm pretty sure the tag that triggered the build can be pulled in here.

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [reply .github/workflows/build-and-publish.yml:16]: That's all reasonable - let's stick with with `ubuntu-latest`. Thanks for the explanation!

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [reply .github/workflows/build-and-publish.yml:8]: Sounds excellent - it would also allow us to include some testing steps at a later point, right?

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [reply .github/workflows/build-and-publish.yml:36]: That all makes sense - thanks.

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [reply .github/workflows/build-and-publish.yml:28]: That makes sense - thanks :)

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [reply .github/workflows/build-and-publish.yml]: That would be great, yeah :)

I'm marking this as "unresolved", since it's not yet implemented.

File: .github/workflows/build-and-publish.yml
Comment (chrivers) [reply .github/workflows/build-and-publish.yml:17]: So `packages` is for uploading the artefact, but `contents` is for creating a release object? It kind of looks like `contents` also includes complete write access to the whole repository, but I guess we can live with that?

**Review by chrivers — CHANGES_REQUESTED** [review]
Lots of stuff well explained - thanks :)

I'd like to see the naming fixed to include the tag (or other descriptive identifier) before we merge this, but otherwise I feel like we are almost there.

File: .github/workflows/build-and-publish.yml
Comment (Cisien) [reply .github/workflows/build-and-publish.yml:17]: Yeah, not ideal. Fortunately, github doesn't allow any ol' random PR access to the GITHUB_TOKEN by default.

File: .github/workflows/build-and-publish.yml
Comment (Cisien) [reply .github/workflows/build-and-publish.yml:25]: I'm going to leave it for now, we an nuke it when the submodule goes away, or just let it stick around in case we need submodules in the future. As you said, it doesn't hurt anything

**Review by chrivers — APPROVED** [review]

**Pull request merged**
